pkg/ddl/partition.go

```diff
@@ -2185,9 +2185,7 @@ func (w *worker) rollbackLikeDropPartition(jobCtx *jobContext, job *model.Job) (
 
 	var dropIndices []*model.IndexInfo
 	for _, indexInfo := range tblInfo.Indices {
-		if indexInfo.Unique &&
-			indexInfo.State == model.StateDeleteReorganization &&
-			tblInfo.Partition.DDLState == model.StateDeleteReorganization {
+		if indexInfo.State == model.StateWriteOnly {
```

Comment:
Why only drop the write-only indexes? What about other states? For example, what will happen when we rollback from the "none -> delete-only" step?

There may be `delete-only` indexes before converting to rolling back job:
```
newIndex := index.Clone()
newIndex.State = model.StateDeleteOnly
newIndex.ID = AllocateIndexID(tblInfo)
```

Reply:
There should only be either `StatePublic` or `StateWriteOnly` so only write-only indexes that should be removed. Also in the follow ups #57114 and #56974 (based on this PR) it will block the rollback [here](https://github.com/pingcap/tidb/pull/56974/files#diff-443bbe2f8987cd40a2d87394088d85420358c25634111278944472b5b840590eL345) and [here](https://github.com/pingcap/tidb/pull/57114/files#diff-c51ad6b394f865592721a3e18d58574176394570e3efcff7703e8bd0c3c5c2efL744) for the last state. One can see this PR a part 1 of 3, where the optimistic path works, then the failure/rollback/cleanup is fixed in part 2 and 3.